experiment/create-build-project.sh

```diff
@@ -0,0 +1,213 @@
+check_getopt() {
+  if getopt --test &>/dev/null; then
+    {
+      echo
+      echo "$SCRIPT_NAME requires the use of the \"GNU getopt\" command."
+      echo
+      echo "brew install gnu-getopt"
+      echo "brew link --force gnu-getopt"
+      exit 1
+    } >&2
+  fi
+}
+
+get_options() {
+  check_getopt
+
+  if opt=$(getopt -o i -l machine-type:,num-nodes:,cluster-name:,zone:,pod-namespace:,interactive -n "$SCRIPT_NAME" -- "$@"); then
+    eval set -- "$opt"
+  else
+    {
+      echo
+      echo "unable to parse options"
+      exit 1
+    } >&2
+  fi
+
+  while [ $# -gt 0 ]; do
+    case "$1" in
+    --machine-type)
+      MACHINE_TYPE="$2"
+      shift 2
+      ;;
+    --num-nodes)
+      NUM_NODES="$2"
+      shift 2
+      ;;
+    --cluster-name)
+      CLUSTER_NAME="$2"
+      shift 2
+      ;;
+    --zone)
+      ZONE="$2"
+      shift 2
+      ;;
+    --pod-namespace)
+      POD_NAMESPACE="$2"
+      shift 2
+      ;;
+    -i | --interactive)
+      INTERACTIVE=1
+      shift
+      ;;
+    --)
+      shift
+      get_params "$@"
+      break
+      ;;
+    *)
+      {
+        echo
+        echo "unknown option: $1"
+        exit 1
+      } >&2
+      ;;
+    esac
+  done
+}
+
+get_params() {
+  if [ $# -eq 0 ]; then
+    {
+      echo
+      echo "project name required"
+      exit 1
+    } >&2
+  fi
+
+  PROJECT_NAME="$1"
+  PROJECT_ID="$(echo "$PROJECT_NAME" | tr '[A-Z]' '[a-z]' | tr ' ' '-')"
+}
+
+# Create GCP project.
+create_project() {
+  gcloud projects create "$PROJECT_ID" --name="$PROJECT_NAME" --folder="$FOLDER_ID"
+}
+
+# Link project to billing account.
+link_billing() {
+  gcloud beta billing projects link "$PROJECT_ID" --billing-account="$BILLING_ACCOUNT"
+}
+
+# Enable Kubernetes Engine API.
+enable_kube_api() {
+  gcloud services enable "container.googleapis.com" --project="$PROJECT_ID"
+}
+
+# Add iam policy bindings.
+add_iam() {
+  for member in "${!IAM_ROLES[@]}"; do
+    gcloud projects add-iam-policy-binding "$PROJECT_ID" --member="$member" --role="${IAM_ROLES[$member]}"
+  done
+}
+
+# Create cluster.
+create_cluster() {
+  gcloud container clusters create "$CLUSTER_NAME" --project="$PROJECT_ID" --zone="$ZONE" --machine-type="$MACHINE_TYPE" --num-nodes="$NUM_NODES"
+}
+
+# Authenticate to cluster.
+get_creds() {
+  gcloud container clusters get-credentials "$CLUSTER_NAME" --project="$PROJECT_ID" --zone="$ZONE"
+}
+
+# Create pod namespace in cluster.
+create_namespace() {
+  kubectl create namespace "$POD_NAMESPACE"
+}
+
+# Restore kube context.
+restore_kube_context() {
+  kubectl config use-context "$CURRENT_CONTEXT" &>/dev/null
+}
+
+main() {
+  get_options "$@"
+
+  # gcloud
+  prompt "Create project" create_project
+  prompt "Link billing account" link_billing
+  prompt "Enable Kubernetes API" enable_kube_api
+  prompt "Add IAM bindings" add_iam
+  prompt "Create cluster" create_cluster
+
+  get_creds
+
+  # kubectl
+  prompt "Create namespace" create_namespace
+}
+
+main "$@"
```

Comment:
Why not make this a go program?

Reply:
I added this to `experiment` because it is not the most robust solution. I do plan to add this to mkbuild-cluster or a similarly named tool. I supposed as @BenTheElder suggested terraform is an option as well. I will get to this soon; I want to use the Gcloud SDK if possible to be the most portable. Let me create add this to the issue I have for improvements to tackle/mkbuild-cluster.

Reply:
[I wouldn't block on using terraform, but it might be nice to align with wg-k8s-infra there and it has served pretty well so far]

Reply:
I am all for terraform! I am not familiar with the syntax, although it seems pretty straightforward. As long as it supports [input variables](https://learn.hashicorp.com/terraform/getting-started/variables.html) (which it appears to) it should work for this.

Reply:
I will look into it.